File: library/src/main/java/com/github/library/model/OAuthResponse.java
package com.github.library.model;


import java.io.IOException;

import com.github.library.helper.general;

import com.github.library.oauth2.OAuthError;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class OAuthResponse {

    private Long expiresAt;
    private String responseBody;
    private Token token;
    private OAuthError error;
    private Response response;
    private boolean jsonParsed;

    public OAuthResponse(Response response) throws IOException {
        this.response = response;
        if (response != null) {
            ResponseBody body = response.body();
            if (body  == null){
                return;
            }
            responseBody = body.string();
            if (general.isJsonResponse(response)) {
                if (response.isSuccessful()) {
                    token = new Token(responseBody);
                    jsonParsed = true;
                    if (token.expires_in != null)
                        expiresAt = (token.expires_in * 1000) + System.currentTimeMillis();
                } else {
                    try {
                        //error = moshi.adapter(OAuthError.class).fromJson(responseBody);
                        jsonParsed = true;
                    } catch (Exception e) {
                        error = new OAuthError(e);
                        jsonParsed = false;

                    }
                }
            }
        }
    }

    public OAuthResponse(Exception e) {
        this.response = null;
        this.error = new OAuthError(e);
    }

    public boolean isSuccessful() {
        return response != null && response.isSuccessful() && jsonParsed;
    }

    public boolean isJsonResponse() {
        return jsonParsed;
    }

    public Integer getCode() {
        return response != null ? response.code(): null;
    }

    public Long getExpiresIn() {
        return token != null ? token.expires_in : null;
    }

    public Long getExpiresAt() {
        return expiresAt;
    }

    public String getTokenType() {
        return token != null ? token.token_type : null;
    }

    public String getRefreshToken() {
        return token != null ? token.refresh_token : null;
    }

    public String getAccessToken() {
        return token != null ? token.access_token : null;
    }

    public String getScope() {
        return token != null ? token.scope : null;
    }

    public String getBody() {
        return responseBody;
    }

    public OAuthError getOAuthError() {
        return error;
    }

    public Response getHttpResponse() {
        return response;
    }

}

File: library/src/main/java/com/github/library/model/OAuthResponse.kt
package com.github.library.model


import com.github.library.helper.General
import com.github.library.oauth2.OAuthError
import okhttp3.Response
import java.io.IOException

class OAuthResponse {

    var expiresAt: Long? = null
        private set
    var body: String? = null
        private set
    private var token: Token? = null
    var oAuthError: OAuthError? = null
        private set
    var httpResponse: Response? = null
        private set
    var isJsonResponse: Boolean = false
        private set

    val isSuccessful: Boolean
        get() = httpResponse != null && httpResponse!!.isSuccessful && isJsonResponse

    val code: Int?
        get() = if (httpResponse != null) httpResponse!!.code() else null

    val expiresIn: Long?
        get() = if (token != null) token!!.expires_in else null

    val tokenType: String?
        get() = if (token != null) token!!.token_type else null

    val refreshToken: String?
        get() = if (token != null) token!!.refresh_token else null

    val accessToken: String?
        get() = if (token != null) token!!.access_token else null

    val scope: String?
        get() = if (token != null) token!!.scope else null

    @Throws(IOException::class)
    constructor(response: Response?) {
        this.httpResponse = response
        if (response != null) {
            val body = response.body() ?: return
            this.body = body.string()
            if (General.isJsonResponse(response)) {
                if (response.isSuccessful) {
                    token = Token(this.body!!)
                    isJsonResponse = true
                    if (token!!.expires_in != null)
                        expiresAt = token!!.expires_in!! * 1000 + System.currentTimeMillis()
                } else {
                    try {
                        isJsonResponse = true
                    } catch (e: Exception) {
                        oAuthError = OAuthError(e)
                        isJsonResponse = false

                    }

                }
            }
        }
    }

    constructor(e: Exception) {
        this.httpResponse = null
        this.oAuthError = OAuthError(e)
    }

}
